### ISSUE TITLE
Pyright does not recognize that pytest.fail always raises an exception

### ISSUE THREAD

**Issue by pszpetkowski** [microsoft/pyright#2744]
**Describe the bug**
It seems like most of the time, pyright correctly finds out that other functions raise an exception, but in case of `pytest.fail` it doesn't work. I've tried recreating it on my own and it seems that the `_with_exception` decorator that is used for the `pytest.fail` function is the problem. I'm not entirely sure why, could be related to the `_WithException` shenanigans.

**To Reproduce**
1. Install pytest.
2. Create a new file with the code example defined below.
3. Receive an error reported just like in the screenshot.

**Expected behavior**
In the example below after `pytest.fail("...")` line, pyright should know that `result` is of type `str`.

**Screenshots or Code**

![Screenshot from 2021-12-21 15-19-59](https://user-images.githubusercontent.com/18597734/146945155-c5b913ba-6579-4d3a-a7ef-362adc4706b3.png)

```
import pytest

def get_result() -> str | None:
    return None


def test_foo():
    result = get_result()
    if result is None:
        pytest.fail("...")

    data = result.title()
    print(data)
```

**VS Code extension or command-line**
VS Code extension through Pylance v2021.12.2.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.199, which I just published. It will also be included in the next release of pylance.
